### PR TITLE
Allow parallelisation of masks

### PR DESCRIPTION
**Pull request by dordsor21**
tested and seems to work fine so ¯\\\_(ツ)\_/¯

### CODE REVIEW

**Comment by dordsor21** [comment]
image
nice

**Comment by aurorasmiles** [comment]
What does this actually do? xD

**Comment by dordsor21** [comment]
The main point of it is this:
https://github.com/IntellectualSites/FastAsyncWorldEdit/pull/636/files#diff-bf597b58c1af88ba4850af6ae9fd27a7L103-L108

It had to be synchronised because a lot of masks are not thread safe. With this PR, masks are cloneable so we can have one mask per thread to account for this.

**Review by N0tMyFaultOG — APPROVED** [review]

**Review by SirYwell — COMMENTED** [review]
I don't like the usage of `Object#clone` here, as the implementations violate the javadocs of the clone method (`x.clone() != x` might be false). I'd prefer something similar to the `Filter#fork` method or just `copy()` (don't know why Mask2D has `copy` and `Mask` has `clone`

**Comment by dordsor21** [comment]
> I don't like the usage of `Object#clone` here, as the implementations violate the javadocs of the clone method (`x.clone() != x` might be false). I'd prefer something similar to the `Filter#fork` method or just `copy()` (don't know why Mask2D has `copy` and `Mask` has `clone`

They had to be different because AlwaysTrue and AlwaysFalse implement both Mask and Mask2D

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import com.sk89q.worldedit.math.transform.Transform;
+import com.sk89q.worldedit.world.block.BlockType;
+
+public class ImageBrushMask extends AbstractExtentMask {
+
+    private final MutableVector3 mutable = new MutableVector3();
+    private final Mask solid;
+    private final int cx;
+    private final int cy;
+    private final int cz;
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java:20]: please rename to chunkX, chunkY, chunkZ

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    private final int width;
+    private final int height;
+    private final ImageBrush.ColorFunction colorFunction;
+    private final EditSession session;
+    private final TextureUtil texture;
+
+    public ImageBrushMask(Mask solid,
+                          int cx,
+                          int cy,
+                          int cz,
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java:34]: chunkX, chunkY, chunkZ

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        this.centerX = centerX;
+        this.centerZ = centerZ;
+        this.width = width;
+        this.height = height;
+        this.colorFunction = colorFunction;
+        this.session = session;
+        this.texture = texture;
+    }
+
+    @Override public boolean test(BlockVector3 vector) {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java:60]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                if (block != null) {
+                    session.setBlock(vector, block.getDefaultState());
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @Override public Mask clone() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/ImageBrushMask.java:101]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/LayerBrushMask.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    public LayerBrushMask(EditSession editSession, RecursiveVisitor visitor, BlockState[] layers, Mask adjacent) {
+        super(editSession);
+        this.editSession = editSession;
+        this.visitor = visitor;
+        this.layers = layers;
+        this.adjacent = adjacent;
+    }
+
+    @Override public boolean test(BlockVector3 pos) {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/LayerBrushMask.java:29]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/StencilBrushMask.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.concurrent.ThreadLocalRandom;
+
+public class StencilBrushMask extends AbstractExtentMask {
+
+    private final MutableVector3 mutable = new MutableVector3();
+    private final EditSession editSession;
+    private final Mask solid;
+    private final int cx;
+    private final int cy;
+    private final int cz;
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/StencilBrushMask.java:23]: chunkX, chunkY, chunkZ

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/StencilBrushMask.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    private final int add;
+    private final int cutoff;
+    private final int maxY;
+    private final Pattern pattern;
+
+    public StencilBrushMask(EditSession editSession,
+                            Mask solid,
+                            int cx,
+                            int cy,
+                            int cz,
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/StencilBrushMask.java:37]: chunkX, chunkY, chunkZ

File: worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/StencilBrushMask.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            }
+            if (val >= 255 || ThreadLocalRandom.current().nextInt(maxY) < val) {
+                editSession.setBlock(vector.getBlockX(), vector.getBlockY(), vector.getBlockZ(), pattern);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    @Override public Mask clone() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/boydti/fawe/object/brush/mask/StencilBrushMask.java:90]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java [hunk 9]
```diff
@@ -113,6 +117,10 @@ public boolean test(BlockVector3 vector) {
             public Mask2D toMask2D() {
                 return mask;
             }
+
+            @Override public Mask clone() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java:121]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java [hunk 10]
```diff
@@ -143,6 +151,16 @@ public Mask tryOr(Mask other) {
             return this;
         }
 
+        // No need to properly clone an always true mask
+        @Override public Mask clone() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java:155]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java [hunk 11]
```diff
@@ -143,6 +151,16 @@ public Mask tryOr(Mask other) {
...
+            return new AlwaysTrue();
+        }
+
+        // No need to properly clone an always true mask
+        @Override public Mask2D copy() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java:160]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java [hunk 12]
```diff
@@ -92,6 +92,10 @@ public static Mask2D negate(final Mask2D mask) {
             public boolean test(BlockVector2 vector) {
                 return !mask.test(vector);
             }
+
+            @Override public Mask2D copy() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java:96]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java [hunk 13]
```diff
@@ -172,6 +190,16 @@ public Mask tryOr(Mask other) {
             return other;
         }
 
+        // No need to properly clone an always false mask
+        @Override public Mask clone() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java:194]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java [hunk 14]
```diff
@@ -172,6 +190,16 @@ public Mask tryOr(Mask other) {
...
+            return new AlwaysFalse();
+        }
+
+        // No need to properly clone an always true mask
+        @Override public Mask2D copy() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/Masks.java:199]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/WallMakeMask.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+public class WallMakeMask implements Mask {
+
+    private final Region region;
+
+    public WallMakeMask(Region region) {
+        this.region = region.clone();
+    }
+
+    @Override public boolean test(BlockVector3 position) {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/WallMakeMask.java:14]: `@Override` should be on its own line.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/WallMakeMask.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+        int x = position.getBlockX();
+        int z = position.getBlockZ();
+        return !region.contains(x, z + 1) || !region.contains(x, z - 1) || !region.contains(x + 1, z) || !region.contains(x - 1, z);
+    }
+
+    @Override public Mask clone() {
```
Comment (MattBDev) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/mask/WallMakeMask.java:20]: `@Override` should be on its own line.

**Review by MattBDev — CHANGES_REQUESTED** [review]

**Review by aurorasmiles — APPROVED** [review]

**Pull request merged**